main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -27,19 +28,154 @@ import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps, Validation}
+        case Expression.ArrayLoad(base, index, _, _, _) => visit(base) ++ visit(index)
+        case Expression.ArrayLength(base, _, _) => visit(base)
+        case Expression.ArrayStore(base, index, elm, _) => visit(base) ++ visit(index) ++ visit(elm)
+        case Expression.ArraySlice(base, beginIndex, endIndex, _, _) => visit(base) ++ visit(beginIndex) ++ visit(endIndex)
+        case Expression.Ref(exp, _, _, _) => visit(exp)
+        case Expression.Deref(exp, _, _, _) => visit(exp)
+        case Expression.Assign(exp1, exp2, _, _, _) => visit(exp1) ++ visit(exp2)
+        case Expression.Existential(_, exp, _) => visit(exp)
+        case Expression.Universal(_, exp, _) => visit(exp)
+        case Expression.Ascribe(exp, _, _, _) => visit(exp)
+        case Expression.Cast(exp, _, _, _) => visit(exp)
+        case Expression.TryCatch(exp, rules, _, _, _) => visit(exp) ++ rules.flatMap(rule => visit(rule.exp))
+        case Expression.InvokeConstructor(_, args, _, _, _) => args.flatMap(visit)
+        case Expression.InvokeMethod(_, exp, args, _, _, _) => visit(exp) ++ args.flatMap(visit)
+        case Expression.InvokeStaticMethod(_, args, _, _, _) => args.flatMap(visit)
+        case Expression.GetField(_, exp, _, _, _) => visit(exp)
+        case Expression.PutField(_, exp1, exp2, _, _, _) => visit(exp1) ++ visit(exp2)
+        case Expression.GetStaticField(_, _, _, _) => Nil
+        case Expression.PutStaticField(_, exp, _, _, _) => visit(exp)
+        case Expression.NewChannel(exp, _, _, _) => visit(exp)
+        case Expression.GetChannel(exp, _, _, _) => visit(exp)
+        case Expression.PutChannel(exp1, exp2, _, _, _) => visit(exp1) ++ visit(exp2)
+        case Expression.SelectChannel(rules, default, _, _, _) => rules.flatMap(rule => visit(rule.chan) ++ visit(rule.exp)) ++ default.toList.flatMap(visit)
+        case Expression.Spawn(exp, _, _, _) => visit(exp)
+        case Expression.Lazy(exp, _, _) => visit(exp)
+        case Expression.Force(exp, _, _, _) => visit(exp)
+        case Expression.FixpointConstraintSet(_, _, _, _) => Nil
+        case Expression.FixpointCompose(exp1, exp2, _, _, _, _) => visit(exp1) ++ (visit(exp2))
+        case Expression.FixpointSolve(exp, _, _, _, _) => visit(exp)
+        case Expression.FixpointProject(_, exp, _, _, _) => visit(exp)
+        case Expression.FixpointEntails(exp1, exp2, _, _, _) => visit(exp1) ++ visit(exp2)
+        case Expression.FixpointFold(_, exp1, exp2, exp3, _, _, _) => visit(exp1) ++ visit(exp2) ++ visit(exp3)
+      }
+
+      visit(defn0.exp)
+    }
+
+    /**
+      * Checks that all signatures in `class0` are used in laws, unless `class0` is marked `lawless`.
+      */
+    def checkLawApplication(class0: TypedAst.Class): Validation[Unit, InstanceError] = class0 match {
+      case TypedAst.Class(_, mod, _, _, _, signatures, laws, _) =>
+        if (mod.isLawless) {
+          ().toSuccess
+        } else {
+          val usedSigs = laws.flatMap(findUsedSigs)
```

Comment:
A flatMap on a list is potentially expensive-- better to fold on a set.

Reply:
Compare code to when using a list:

```
def checkLawApplication(class0: TypedAst.Class): List[InstanceError] = class0 match {
  case TypedAst.Class(_, mod, _, _, _, signatures, laws, _) if mod.isLawless => Nil
  case TypedAst.Class(_, mod, _, _, _, signatures, laws, _) =>
    val usedSigs = laws.foldLeft(Set.empty[Symbol.SigSym]) {
      case (acc, law) => acc ++ findUsedSigs(law)
    }
    usedSigs.map(sig => InstanceError.UnlawfulSignature(sig.sym, sig.loc))
}
```

Certainly shorter, but perhaps less future proof?

Reply:
I was avoiding sets since you previously mentioned preferring lists for determinism. So I guess it's not a big deal if it's local?

Reply:
I think the point is that there is a big difference between having set semantics in an AST and very locally computing with a few sets. Also symbols have very well-define equality so it is clear what the meaning of the set is.

Reply:
Thanks for calling it out :)